### PR TITLE
Jsbsim issue 312 proposal 1

### LINKED ISSUE #312: Replace remaining 15 exit calls from FGXMLElement.cpp with matching exceptions
Libraries calling `exit` are less well to integrate. Commit 422910b6843baed4099f825c54d84d1e4df8fa0e already replaced the `exit` call in `FGXMLParser.cpp` with a bare `throw`.

I suggest to replace the 15 occurrences of `exit` calls in `FGXMLElement.cpp` with throws of matching exceptions from `std::exception` as a service for library users. An example would be [std::domain_error](https://en.cppreference.com/w/cpp/error/domain_error) in case the value is specifically out of the expected domain. Another would be [std::invalid_argument](https://en.cppreference.com/w/cpp/error/invalid_argument) if value is more broadly unexpected/invalid.

If no one objects I may very well provide a minimal invasive pull request this week.

### PR DESCRIPTION
**Pull request by sthagen**
This is my first proposal assembling what I understood from the discussions would be a non breaking change enhancing the library posture of [JSBSim](https://github.com/JSBSim-Team/jsbsim). In case it is accepted this set of changes closes #312 

@bcoconni and @seanmcleod: Any feedback very welcome!

## Why change?
Be nice as a library and let users decide if an XML parsing error is fatal or not (because only the user knows)

## What does it accomplish?
All `exit(...)` calls could be transformed into exception handling use cases (we throw ...).

## How does it accomplish the transformation?
Use cases mapping:
- emptiness where a value was expected (`length_error`)
- conversion failures because of non-existing conversion rules (`invalid_argument`)
- conversion failures because of value errors (`domain_error`)
- conversion failures because of unknown errors (`invalid_argument`)
- cardinality errors e.g. single value expected, but no value or multiple values found (`length_error`)
- access errors when requesting a non-existing element (`length_error`)

In most cases the error message already built was reused as `what` argument to the exception.

A few cases build elaborate output per multiple values found and these were not translated into the exception argument (to keep these "reason" parameter as a "single line with no line end" string).

Two new standard library headers are introduced and the rationale is noted as comments in place.
~~The change also introduces a C++11 helper function that accepts any heterogeneous arguments to collate into an error string.~~

~~A wrapper comment section has been added to mimic the layout style of the time the source was invented 😉~~

## Observable Behavior Changes

1. In case the execution enters the changed "error handling branches" exceptions are thrown instead of killing the whole process.
2. Users watching the log via std err will now have repeated cause statements and stack traces in addition to the still written error messages.

Rationale for 2.: The error messages may have been considered part of the public API by some users so this pull request does not touch those.

### Additional Notes
- ~~When migrating the code base from C++11 to C++17 in the future the helper function can be greatly simplified (as noted in comments in the function).~~
- In some places the existing implementation does not exit but instead returns some marker like value and writes a log message to the standard error stream. These places might benefit from revisiting if such silent (for users not watching the console logs) continuation does not meet the expectation of most library users.
- At least three of the error handling groups could be handled in a central place but for now were kept as is to have a minimal invasive change.
- We could consider removing the log messages when throwing or keeping after surveying use cases of the library?

### CODE REVIEW

File: src/input_output/FGXMLElement.cpp [hunk 1]
```diff
@@ -42,6 +44,19 @@ namespace JSBSim {
...
+
+template <typename... T>
+std::string _concat_to_string(T&&... ts) {
+    std::stringstream s;
+    // in C++17 below statement can be simply: (s << ... << std::forward<T>(ts));
+    int dummy[] = { 0, ((s << std::forward<T>(ts)), 0)... };  // C++11 version with perfect forwarding
+    static_cast<void>(dummy); // Avoid warning for unused variable in C++11 (can drop in C++17)
+    return s.str();
+}
+
```
[the remark this answers is not in the document]
Comment (seanmcleod) [reply src/input_output/FGXMLElement.cpp:59]: I agree with @bcoconni without the name of the function and a mention of it in your overall description it wasn't clear to what the code was doing, and even with that I don't completely follow how it works.

[the remark this answers is not in the document]
Comment (sthagen) [reply src/input_output/FGXMLElement.cpp:59]: Thank you very much for the readable resolution.

File: src/input_output/FGXMLElement.cpp [hunk 2]
```diff
@@ -334,22 +351,28 @@ double Element::GetDataAsNumber(void)
...
     }
 
     return number;
   } else if (data_lines.size() == 0) {
-    cerr << ReadFrom() << "Expected numeric value, but got no data" << endl;
-    exit(-1);
+    string const log_message{_concat_to_string(
+      ReadFrom(), "Expected numeric value, but got no data")};
+    cerr << log_message << endl;
+    throw length_error(log_message);
```
[the remark this answers is not in the document]
Comment (sthagen) [reply src/input_output/FGXMLElement.cpp:365]: Will do.

File: src/input_output/FGXMLElement.cpp [hunk 3]
```diff
@@ -528,21 +554,26 @@ double Element::FindElementValueAsNumberConvertFromTo( const string& el,
...
     if (convert[supplied_units].find(target_units) == convert[supplied_units].end()) {
-      cerr << element->ReadFrom() << "Supplied unit: \""
-           << supplied_units << "\" cannot be converted to " << target_units
-           << endl;
-      exit(-1);
+      string const log_message{_concat_to_string(
+        element->ReadFrom(), "Supplied unit: \"", supplied_units,
+        "\" cannot be converted to ", target_units)};
+      cerr << log_message << endl;
+      throw invalid_argument(log_message);
```
[the remark this answers is not in the document]
Comment (sthagen) [reply src/input_output/FGXMLElement.cpp:576]: :+1:

File: src/input_output/FGXMLElement.cpp [hunk 4]
```diff
@@ -567,15 +598,17 @@ FGColumnVector3 Element::FindElementTripletConvertTo( const string& target_units
...
     if (convert[supplied_units].find(target_units) == convert[supplied_units].end()) {
-      cerr << ReadFrom() << "Supplied unit: \""
-           << supplied_units << "\" cannot be converted to " << target_units
-           << endl;
-      exit(-1);
+      string const log_message{_concat_to_string(
+        ReadFrom(), "Supplied unit: \"", supplied_units, "\" cannot be converted to ",
+        target_units)};
+      cerr << log_message << endl;
+      throw invalid_argument(log_message);
```
[the remark this answers is not in the document]
Comment (sthagen) [reply src/input_output/FGXMLElement.cpp:611]: 👍

**Comment by sthagen** [comment]
I will create the remaining changes tomorrow. Thank you again for the fast and helpful feedback @bcoconni and @seanmcleod

**Comment by sthagen** [comment]
Embedded review feedback (as understood) and updated the PR description accordingly.

**Comment by bcoconni** [comment]
@sthagen, just a couple of thoughts about your additional notes:
>In some places the existing implementation does not exit but instead returns some marker like value and writes a log message to the standard error stream. These places might benefit from revisiting if such silent (for users not watching the console logs) continuation does not meet the expectation of most library users.

That is indeed something that needs some care: JSBSim is 20+ years old and inconsistencies have accumulated over the years. It is worthwhile to revisit the code from time to time to ensure that a minimum level of consistency is maintained. Suggestions are welcome !
> At least three of the error handling groups could be handled in a central place but for now were kept as is to have a minimal invasive change.

Everything that alleviate the maintenance burden is welcome as well !
> We could consider removing the log messages when throwing or keeping after surveying use cases of the library?

JSBSim is a command line tool at its core. Even if it can be used as a library or as a Python module, JSBSim will continue to be designed as a command line tool and, as a consequence, it must (at least optionally) log its activities in the console. For example, most aircraft developers for FlightGear use the console messages from JSBSim to debug their model.

I agree however that our log system is extremely basic since it basically sums up to choosing between outputting to `cout` or `cerr`. I have toyed with the idea of using FlightGear log class `SG_LOG` which is much more versatile but as always I have much more ideas than time to implement them 😞

**Comment by sthagen** [comment]
@bcoconni thank you very much for the prompt & thorough reviews. I am happy to contribute minimally invasive changes when they are welcome (as it seems ❤️) and also enhance the library, operational, and maintenance postures. It was a pleasure!

**Comment by jonsberndt** [comment]
Excellent enhancements.

**Pull request merged**